jax/_src/lax/lax.py

```diff
@@ -4681,53 +4710,97 @@ def _scatter_jvp(primals, tangents, *, update_jaxpr, update_consts,
   partial(_scatter_batching_rule, scatter))
 
 
-def _reduce_shape_rule(operand, init_value, *, computation, jaxpr, consts,
-                       dimensions):
-  return tuple(np.delete(operand.shape, dimensions))
+def _reduce_abstract_eval_rule(*avals, computation, jaxpr, consts, dimensions):
+  assert all(isinstance(aval, UnshapedArray) for aval in avals), avals
+  avals = tuple(raise_to_shaped(aval) for aval in avals)
+  operand_avals, init_value_avals = split_list(avals, [len(avals) // 2])
+  if any(aval.shape != () for aval in init_value_avals):
+    init_value_shapes = [a.shape for a in init_value_avals]
+    raise ValueError(f'Found non-scalar init_value: {init_value_shapes}')
+  least_specialized = _max(
+      map(type, avals), key=operator.attrgetter('array_abstraction_level'))
+  if least_specialized is ConcreteArray:
+    return safe_map(ConcreteArray,
+                    reduce_p.impl(*[x.val for x in avals],computation=computation,
+                                  jaxpr=jaxpr, consts=consts, dimensions=dimensions))
+  elif least_specialized is ShapedArray:
+    return [
+        ShapedArray(tuple(np.delete(op_aval.shape, dimensions)),
+                    dtypes.canonicalize_dtype(op_aval.dtype))
+        for op_aval in operand_avals
+    ]
+  elif least_specialized is UnshapedArray:
+    return [
+        UnshapedArray(dtypes.canonicalize_dtype(in_aval.dtype))
+        for in_aval in init_value_avals
+    ]
+  else:
+    raise TypeError(avals, least_specialized)
 
-def _reduce_translation_rule(c, operand, init_value, *, computation, jaxpr,
-                             consts, dimensions):
-  xla_computation = _reduction_computation(c, jaxpr, consts, init_value)
-  return xops.Reduce(c, [operand], [init_value], xla_computation, dimensions)
 
-def _reduce_batch_rule(batched_args, batch_dims, *, computation, jaxpr, consts,
-                       dimensions):
-  operand, init_value = batched_args
-  operand_bdim, init_value_bdim = batch_dims
-  if init_value_bdim is None:
-    assert operand_bdim is not None
+def _reduce_translation_rule(c, *values, computation, jaxpr,
+                             consts, dimensions):
+  operands, init_values = split_list(values, [len(values) // 2])
+  xla_computation = _reduction_computation(c, jaxpr, consts, init_values, singleton=False)
+  return xops.Reduce(c, operands, init_values, xla_computation, dimensions)
+
+def _reduce_batch_rule(batched_args, batch_dims, *, computation, jaxpr,
+                       consts, dimensions):
+  num_operands = len(batched_args) // 2
+  operands, init_values = split_list(batched_args, [num_operands])
+  operand_bdims, init_value_bdims = split_list(batch_dims, [num_operands])
+  if all(init_value_bdim is None for init_value_bdim in init_value_bdims):
+    assert all(operand_bdim is not None for operand_bdim in operand_bdims)
```

Comment:
It's not clear to me that these two assertions always hold. I think you actually need to handle all of these cases.

Reply:
I added a TODO.